### PR TITLE
[SPARK-6257] [PySpark] [MLlib] MLlib API missing items in Recommendation

### PR DESCRIPTION
**Pull request by MechCoder**
Adds

rank, recommendUsers and RecommendProducts to MatrixFactorizationModel in PySpark.

### CODE REVIEW

**Comment by MechCoder** [comment]
@jkbradkey

Does it make sense to have things like "setXXX" in Python especially since the attributes are always accessible and especially in this case since the recommendation is trained using methods of the static object ALS?

**Comment by MechCoder** [comment]
Also the release deadline means that PR's would still be reviewed right? It is just that they would enter the 1.5 branch instead of the 1.4?

File: python/pyspark/mllib/recommendation.py [hunk 1]
```diff
@@ -105,9 +112,15 @@ class MatrixFactorizationModel(JavaModelWrapper, JavaSaveable, JavaLoader):
     ...     pass
     """
     def predict(self, user, product):
+        """
+        Predicts rating for a given user and product.
```
Comment (mengxr) [inline python/pyspark/mllib/recommendation.py]: `a given user and product pair` or `the given user and product`

File: python/pyspark/mllib/recommendation.py [hunk 2]
```diff
@@ -105,9 +112,15 @@ class MatrixFactorizationModel(JavaModelWrapper, JavaSaveable, JavaLoader):
...
+        """
         return self._java_model.predict(int(user), int(product))
 
     def predictAll(self, user_product):
+        """
+        Returns a list of predicted Ratings for a user and many products.
```
Comment (mengxr) [inline python/pyspark/mllib/recommendation.py]: `predicted ratings for input user and product pairs`. This works for more than one user.

File: python/pyspark/mllib/recommendation.py [hunk 3]
```diff
@@ -115,11 +128,39 @@ def predictAll(self, user_product):
         return self.call("predict", user_product)
 
     def userFeatures(self):
+        """
+        Returns a coupled RDD, where the first element is the user and the
```
Comment (mengxr) [inline python/pyspark/mllib/recommendation.py]: `coupled` -> `pair`

File: python/pyspark/mllib/recommendation.py [hunk 4]
```diff
@@ -115,11 +128,39 @@ def predictAll(self, user_product):
...
+        second is an array of features corresponding to that user.
+        """
         return self.call("getUserFeatures").mapValues(lambda v: array.array('d', v))
 
     def productFeatures(self):
+        """
+        Returns a coupled RDD, where the first element is the product and the
```
Comment (mengxr) [inline python/pyspark/mllib/recommendation.py]: `coupled` -> `pair`

File: python/pyspark/mllib/recommendation.py [hunk 5]
```diff
@@ -115,11 +128,39 @@ def predictAll(self, user_product):
...
+        second is an array of features corresponding to that product.
+        """
         return self.call("getProductFeatures").mapValues(lambda v: array.array('d', v))
 
+    def recommendUsers(self, product, num):
+        """
+        Recommends the top "num" number of products for a given user.
+        This is done by returning a tuple of Rating objects, the second id (product)
+        being constant and sorted according to the rating.
```
Comment (mengxr) [inline python/pyspark/mllib/recommendation.py]: `Recommends the top "num" number of products for a given user and returns a list or Rating objects sorted by the predicted rating in descending order.`

**Comment by MechCoder** [comment]
@mengxr fixed!

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

**Comment by MechCoder** [comment]
Thanks for the quick comments and merge. Would you be able to have a look at this (https://github.com/apache/spark/pull/5775) too?

**Pull request closed without merging**